unidist/core/backends/mpi/core/monitor.py

```diff
@@ -44,6 +44,105 @@ def increment(self):
         self.task_counter += 1
 
 
+class DataTracker:
+    """
+    Class that keeps track of all completed data_ids.
+
+    """
+
+    __instance = None
+
+    def __init__(self):
+        self.completed_data_ids = set()
+
+    @classmethod
+    def get_instance(cls):
+        """
+        Get instance of ``WaitHandler``.
+
+        Returns
+        -------
+        WaitHandler
+        """
+        if cls.__instance is None:
+            cls.__instance = DataTracker()
+        return cls.__instance
+
+    def add_to_completed(self, data_ids):
+        self.completed_data_ids.update(data_ids)
+
+
+class WaitHandler:
+    """
+    Class that handles wait requests.
+
+    """
+
+    __instance = None
+
+    def __init__(self):
+        self.completed_data_ids = set()
+        self.waited_data_ids = []
+        self.ready = []
+        self.num_returns = 0
+
+    @classmethod
+    def get_instance(cls):
+        """
+        Get instance of ``WaitHandler``.
+
+        Returns
+        -------
+        WaitHandler
+        """
+        if cls.__instance is None:
+            cls.__instance = WaitHandler()
+        return cls.__instance
+
+    def add_wait_request(self, waited_data_ids, num_returns):
+        """
+        Add a wait request for a list of data_ids and number of data_ids to be awaited.
+
+        Parameters
+        ----------
+        waited_data_ids : List[unidist.core.backends.common.data_id.DataID]
+            List of data_ids to be awaited
+        num_returns : int,
+            The number of ``DataID``-s that should be returned as ready.
+        """
+
+        self.waited_data_ids = waited_data_ids
+        self.num_returns = num_returns
+
+    def process_wait_requests(self):
+        """
+        Check if wait requests are pending to be processed.
+
+        Process pending wait requests and send the data_ids if number of data_ids that are
+        ready are equal to the num_returns.
+
+        """
+        i = 0
+        if self.waited_data_ids:
+            while i < len(self.waited_data_ids):
+                data_id = self.waited_data_ids[i]
+                data = {"ready": self.ready, "not_ready": self.waited_data_ids}
+
+                if data_id in DataTracker.get_instance().completed_data_ids:
+                    self.waited_data_ids.remove(data_id)
+                    self.ready.append(data_id)
+                else:
+                    i = i + 1
+                if self.num_returns == len(self.ready):
```

Comment:
What if ready is greater than num_returns? In this case we should count up ready num_returns and add the rest of ready to non ready.